src/core/utils/components/VirtualScroll.tsx

```diff
@@ -0,0 +1,249 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+
+const getScrollableParent = (element: HTMLElement | null): HTMLElement => {
+  if (!element || element === document.body) {
+    return document.body;
+  }
+
+  return isElementScrollable(element)
+    ? element
+    : getScrollableParent(element.parentElement);
+};
+
+const isElementScrollable = (element: HTMLElement) => {
+  return /(auto|scroll|overlay)/.test(
+    getElementStyle(element, 'overflow') +
+      getElementStyle(element, 'overflow-y'),
+  );
+};
+
+const getElementStyle = (element: HTMLElement, prop: string) => {
+  return getComputedStyle(element, null).getPropertyValue(prop);
+};
+
+const getElementHeight = (element: HTMLElement | undefined) => {
+  return !!element ? element.getBoundingClientRect().height : 0;
+};
+
+const getCountOfNodesInHeight = (
+  childHeight: number,
+  totalHeight: number,
+  startIndex = 0,
+) => {
+  let i = startIndex;
+  let sum = 0;
+  while (sum < totalHeight) {
+    sum += childHeight;
+    ++i;
+  }
+  i = sum > totalHeight ? i - 1 : i;
+  return i - startIndex;
+};
+
+const getTranslateValue = (childHeight: number, startIndex: number) => {
+  let sum = 0;
+  let i = 0;
+  while (i < startIndex) {
+    sum += childHeight;
+    ++i;
+  }
+  return sum;
+};
+
+const getVisibleNodeCount = (
+  childHeight: number,
+  startIndex: number,
+  childrenLength: number,
+  scrollContainer: HTMLElement,
+) => {
+  return Math.min(
+    childrenLength - startIndex,
+    getCountOfNodesInHeight(
+      childHeight,
+      getElementHeight(scrollContainer),
+      startIndex,
+    ),
+  );
+};
+
+export type VirtualScrollProps = {
+  /**
+   * Length of the items to virtualize.
+   */
+  itemsLength: number;
+  /**
+   * Single item render function, which gives index of the item (0 based) in the data array
+   * and expects to get the JSX of that element to render.
+   * Recommended to memoize the reference of the function.
+   */
+  itemRenderer: (index: number) => JSX.Element;
+  /**
+   * Number of items to be rendered at the end.
+   * Not recommended to go lower than the visible items in viewport.
+   * @default 20
```

Comment:
20 seems like a big number

Reply:
What is your suggestion? My thought behind it was that we are fine to render even 50 or 100 rows, it should not influence performance. And having 20 in front might save sometimes from unexpected bugs on scroll...

Reply:
Oh, I thought it was rendering 20 before and 20 after. But it only adds 20 at the bottom. Why not on the both ends?

Reply:
why both? Do you see emptiness at the top whiles scrolling or smth? usually its done when you have virtual scroll with fake top and bottom elements with their heights. This solution I dont think requires it

Reply:
I thought you need additional items on top and on the bottom as you mentioned yourself "save sometimes from unexpected bugs on scroll" when you scroll up or down. Also, try tabbing to previous items, you can't because there is none on the top of visible list.

Reply:
Well, I just needed on one side for the initial calculations (where visible nodes are 0, so I am using only those from additionalItems...
So what you are saying.. I should add that same amount on both top and bottom then...?

Reply:
I would say to split it maybe or reduce default value in half.

Reply:
Changed to 10 and rendering from both sides. Will update description too, which now I realized forgot..

Reply:
done